README.md

## What is mc3p?

mc3p (short for Minecraft Protocol-Parsing Proxy) is a Minecraft proxy
server. With mc3p, you can create programs (mc3p plugins) that examine
and modify messages sent between the Minecraft client and server without
writing any multi-threaded or network-related code.

## Installation.

mc3p requires Python 2.7. The easiest way to install mc3p is with
easy_install from the setuptools project. If you don't have setuptools
installed, first follow the installation instructions on
[this page](http://pypi.python.org/pypi/setuptools).

With setuptools installed, installing mc3p is as simple as:

    $ easy_install http://mattmcgill.com/mc3p/mc3p-0.2pre-py2.7.egg

Note that, depending on your system configuration, you may have to run
the command with root privileges.

### Installing from source

To install from source, just clone the GitHub repository. You can then run
mc3p directly from the top-level directory of the repository, or install
it in 'development' mode with

    python setup.py develop

If mc3p is installed in 'development' mode, you can uninstall it with

    python setup.py develop --uninstall

## Running mc3p.

To start an mc3p server that listens on port 80 and forwards connections
to a Minecraft server:

    $ python -m mc3p.proxy -p 80 <server>

Within your Minecraft client, you can then connect to <server> through
mc3p using the server address 'localhost:80'. However, to do anything useful
you must enable some plugins.

## Using mc3p plugins.

An mc3p plugin has complete control over all the messages that pass between
the Minecraft client and server. By manipulating messages sent between client
and server, you can add useful functionality without modifying client or server
code.

To run a plugin, you must enable it with the --plugin <name> option when you start mc3p.
All enabled plugins are initialized after a client successfully connects to a server.
For example, to run the mute example plugin that comes with mc3p:

    $ python -m mc3p.proxy --plugin 'mc3p.plugin.mute' <server>

Some plugins accept arguments that modify their behavior. To pass arguments
to a plugin, enclose them in parentheses following the plugin's name, like so:
--plugin '<name>(<arguments>)'. Be sure to use quotes, or escape the parentheses
as required by your shell.

    $ python -m mc3p.proxy --plugin '<plugin>(<arguments>)' <server>

## A Plugin Example: mute

The 'mute' plugin is provided as a simple example of mc3p's flexibility.
This plugin allows a player to mute chat messages from selected players on a
server. It requires no modification to either the Minecraft client or server.

Give it a try: start mc3p with the 'mute' plugin enabled:

    $ python -m mc3p.proxy --plugin 'mc3p.plugin.mute' your.favorite.server.com

You can now mute a player by typing '/mute NAME' in chat,
and unmute them with '/unmute NAME'. You can display muted players with '/muted'.

The plugin works by intercepting all Minecraft chat messages, and silently
discarding those sent by muted players.

Now take a look at the source code for the 'mute' plugin:

    from mc3p.plugins import MC3Plugin, msghdlr

    class MutePlugin(MC3Plugin):
        """Lets the client mute players, hiding their chat messages.
        
        The client controls the plugin with chat commands:
            /mute NAME      Hide all messages from player NAME.
            /unmute NAME    Allow messages from player NAME.
            /muted          Show the list of currently muted players.
        """
        def init(self, args):
            self.muted_set = set() # Set of muted player names.

        def send_chat(self, chat_msg):
            """Send a chat message to the client."""
            self.to_client({'msgtype': 0x03, 'chat_msg': chat_msg})

        def mute(self, player_name):
            self.muted_set.add(player_name)
            self.send_chat('Muted %s' % player_name)

        def unmute(self, player_name):
            if player_name in self.muted_set:
                self.muted_set.remove(player_name)
                self.send_chat('Unmuted %s' % player_name)
            else:
                self.send_chat('%s is not muted' % player_name)

        def muted(self):
            self.send_chat('Currently muted: %s' % ', '.join(self.muted_set))

        @msghdlr(0x03)
        def handle_chat(self, msg, source):
            txt = msg['chat_msg']
            if source == 'client':
                # Handle mute commands
                if txt.startswith('/mute '):     self.mute(txt[len('/mute '):])
                elif txt.startswith('/unmute '): self.unmute(txt[len('/unmute '):])
                elif txt == '/muted':            self.muted()
                else: return True # Forward all other chat messages.

                return False # Drop mute plugin commands.
            else:
                # Drop messages containing the string <NAME>, where NAME is a muted player name.
                return not any(txt.startswith('<%s>' % name) for name in self.muted_set)

Every mc3p plugin is a Python module that contains a single *plugin class*, a
subclass of MC3Plugin. A plugin must contain *exactly* one plugin class;
mc3p will print an error if multiple sub-classes of MC3Plugin are found.

Once a client successfully connects to a server through the proxy, mc3p
creates an instance of the plugin class of each enabled plugin, and calls its
'init' method. Plugin classes should override 'init' to perform plugin-specific
set-up. If the plugin was enabled with an argument string, that string is passed in
the 'args' parameter; otherwise, 'args' is None.

A plugin class registers a *message handler* for every message type it wishes
to receive. To register a method of a plugin class as a message handler, decorate
it with '@msghdlr'. The '@msghdlr' decorator takes one or more message types
as arguments. Each message handler should take two arguments (in addition to 'self'):

* 'msg', a dictionary representing the message, and
* 'source', which indicates the sender of the message; source is either
  'client' or 'server'.

The 'msg' dictionary always maps the key 'msgtype' to the message's type, a
number between 0 and 0xFF. If
your message handler is registered for multiple types, you can determine the type
of a given message by checking `msg['msgtype']`. The other key-value pairs
in the 'msg' dictionary depend on the specific message type. See
[messages.py](https://github.com/mmcgill/mc3p/blob/master/mc3p/messages.py)
for a definition of the keys associated with each message type.

A message handler returns a boolean value indicating whether the message should
be forwarded to its destination. A return value of True forwards the message,
while a return value of False silently drops it. The message handler may also
modify the message by changing the values of the 'msg' dictionary, and
returning True.

The mute plugin registers the 'handle_chat' method as a message handler for
messages of type '0x03', which represent chat messages. If the chat message
is sent from the client, we check to see if it is a command to the mute plugin.
If so, then we process it, and drop it by returning False. If not, we forward it
by returning True. If the chat message is from the server, then we forward it
if it was not sent by a currently blocked player.

Along with modifying or dropping messages, a plugin can create new messages
by passing a 'msg' dictionary with a 'msgtype' and all relevant key-value pairs
to the 'to_client' or 'to_server methods, which are defined in the MC3Plugin class.

The mute plugin uses the 'to_client' method to inject chat messages that indicate
the result of each command issued by the user. Note that since these messages
are sent to the client, and not the server, they are not visible to any other
user on the server.



CHANGELOG
0.2pre
- support for protocol versions 17-21 (Up through 1.9pre5)
- Windows now supported (tested with CPython 2.7)
- various bug fixes

0.1pre (Initial release)
- support for protocol version 17 (Minecraft Beta 1.8)
- basic plugin support
- linux/OS X only



mcproto.py
from parsing import *

cli_msgs = [None] * 256
srv_msgs = [None] * 256

cli_msgs[0x00] = \
srv_msgs[0x00] = defmsg(0x00,"Keep Alive",[
    ('id', MC_int)])

cli_msgs[0x01] = defmsg(0x01,"Login Request",[
    ('proto_version',MC_int),
    ('username',MC_string),
    ('map_seed',MC_long),
    ('nu1', MC_int),
    ('nu2', MC_byte),
    ('nu3', MC_byte),
    ('nu4', MC_unsigned_byte),
    ('nu5', MC_unsigned_byte)])
srv_msgs[0x01] = defmsg(0x01,"Login Response",[
    ('eid',MC_int),
    ('reserved',MC_string),
    ('map_seed',MC_long),
    ('server_mode', MC_int),
    ('dimension', MC_byte),
    ('difficulty', MC_byte),
    ('world_height', MC_unsigned_byte),
    ('max_players', MC_unsigned_byte)])

cli_msgs[0x02] = defmsg(0x02,"Handshake",[
    ('username',MC_string)])
srv_msgs[0x02] = defmsg(0x02, "Handshake", [
    ('hash',MC_string)])

cli_msgs[0x03] = \
srv_msgs[0x03] = defmsg(0x03, "Chat",[
    ('chat_msg',MC_string)])

srv_msgs[0x04] = defmsg(0x04, "Time", [
    ('time',MC_long)])

cli_msgs[0x05] = \
srv_msgs[0x05] = defmsg(0x05, "Entity Equipment Spawn",[
    ('eid',MC_int),
    ('slot',MC_short),
    ('item_id',MC_short),
    ('unknown',MC_short)])

srv_msgs[0x06] = defmsg(0x06, "Spawn position",[
    ('x',MC_int),
    ('y',MC_int),
    ('z',MC_int)])

cli_msgs[0x07] = defmsg(0x07, "Use entity", [
    ('eid',MC_int),
    ('target_eid',MC_int),
    ('left_click',MC_bool)])

srv_msgs[0x08] = defmsg(0x08, "Update health", [
    ('health',MC_short),
    ('food', MC_short),
    ('food_saturation', MC_float)])

cli_msgs[0x09] = \
srv_msgs[0x09] = defmsg(0x09, "Respawn", [
    ('world', MC_byte),
    ('difficulty', MC_byte),
    ('mode', MC_byte),
    ('world_height', MC_short),
    ('map_seed', MC_long)])

cli_msgs[0x0a] = defmsg(0x0a, "Player state", [
    ('on_ground',MC_bool)])

cli_msgs[0x0b] = \
srv_msgs[0x0b] = defmsg(0x0b, "Player position", [
    ('x',MC_double),
    ('y',MC_double),
    ('stance',MC_double),
    ('z',MC_double),
    ('on_ground',MC_bool)])

cli_msgs[0x0c] = defmsg(0x0c, "Player look", [
    ('yaw',MC_float),
    ('pitch',MC_float),
    ('on_ground',MC_bool)])

# Note the difference in ordering of 'stance'!
cli_msgs[0x0d] = defmsg(0x0d, "Player position and look",[
    ('x',MC_double),
    ('y',MC_double),
    ('stance',MC_double),
    ('z',MC_double),
    ('yaw',MC_float),
    ('pitch',MC_float),
    ('on_ground',MC_bool)])
srv_msgs[0x0d] = defmsg(0x0d, "Player position and look", [
    ('x',MC_double),
    ('stance',MC_double),
    ('y',MC_double),
    ('z',MC_double),
    ('yaw',MC_float),
    ('pitch',MC_float),
    ('on_ground',MC_bool)])

cli_msgs[0x0e] = \
srv_msgs[0x0e] = defmsg(0x0e, "Digging", [
    ('status',MC_byte),
    ('x',MC_int),
    ('y',MC_byte),
    ('z',MC_int),
    ('face',MC_byte)])

cli_msgs[0x0f] = \
srv_msgs[0x0f] = defmsg(0x0f, "Block placement", [
    ('x',MC_int),
    ('y',MC_byte),
    ('z',MC_int),
    ('dir',MC_byte),
    ('details',MC_item_details)])

cli_msgs[0x10] = \
srv_msgs[0x10] = defmsg(0x10, "Held item selection",[
    ('slot_id', MC_short)])

srv_msgs[0x11] = defmsg(0x11, "Use bed", [
    ('eid', MC_int),
    ('in_bed', MC_bool),
    ('x', MC_int),
    ('y', MC_byte),
    ('z', MC_int)])

cli_msgs[0x12] = \
srv_msgs[0x12] = defmsg(0x12, "Change animation",[
    ('eid',MC_int),
    ('animation',MC_byte)])

cli_msgs[0x13] = \
srv_msgs[0x13] = defmsg(0x13, "Entity action", [
    ('eid',MC_int),
    ('action', MC_byte)])

srv_msgs[0x14] = defmsg(0x14, "Entity spawn", [
    ('eid', MC_int),
    ('name', MC_string),
    ('x', MC_int),
    ('y', MC_int),
    ('z', MC_int),
    ('rotation', MC_byte),
    ('pitch', MC_byte),
    ('curr_item', MC_short)])

cli_msgs[0x15] = \
srv_msgs[0x15] = defmsg(0x15, "Pickup spawn", [
    ('eid',MC_int),
    ('item',MC_short),
    ('count',MC_byte),
    ('data',MC_short),
    ('x',MC_int),
    ('y',MC_int),
    ('z',MC_int),
    ('rotation',MC_byte),
    ('pitch',MC_byte),
    ('roll',MC_byte)])

srv_msgs[0x16] = defmsg(0x16, "Collect item", [
    ('item_eid',MC_int),
    ('collector_eid',MC_int)])

srv_msgs[0x17] = defmsg(0x17, "Add vehicle/object", [
    ('eid',MC_int),
    ('type',MC_byte),
    ('x',MC_int),
    ('y',MC_int),
    ('z',MC_int),
    ('fireball_data',MC_fireball_data)])

srv_msgs[0x18] = defmsg(0x18, "Mob spawn", [
    ('eid',MC_int),
    ('mob_type',MC_byte),
    ('x',MC_int),
    ('y',MC_int),
    ('z',MC_int),
    ('yaw',MC_byte),
    ('pitch',MC_byte),
    ('metadata',parse_metadata)])

srv_msgs[0x19] = defmsg(0x19, "Painting", [
    ('eid', MC_int),
    ('title', MC_string),
    ('x', MC_int),
    ('y', MC_int),
    ('z', MC_int),
    ('type', MC_int)])

srv_msgs[0x1a] = defmsg(0x1a, "Experience orb", [
    ('eid', MC_int),
    ('x', MC_int),
    ('y', MC_int),
    ('z', MC_int),
    ('count', MC_short)])

cli_msgs[0x1b] = \
srv_msgs[0x1b] = defmsg(0x1b, "???", [
    ('d1', MC_float),
    ('d2', MC_float),
    ('d3', MC_float),
    ('d4', MC_float),
    ('d5', MC_bool),
    ('d6', MC_bool)])

cli_msgs[0x1c] = \
srv_msgs[0x1c] = defmsg(0x1c, "Entity velocity", [
    ('eid',MC_int),
    ('vel_x',MC_short),
    ('vel_y',MC_short),
    ('vel_z',MC_short)])

srv_msgs[0x1d] = defmsg(0x1d, "Destroy entity", [
    ('eid',MC_int)])

srv_msgs[0x1e] = defmsg(0x1e, "Entity", [
    ('eid', MC_int)])

srv_msgs[0x1f] = defmsg(0x1f, "Entity relative move", [
    ('eid',MC_int),
    ('dx',MC_byte),
    ('dy',MC_byte),
    ('dz',MC_byte)])

srv_msgs[0x20] = defmsg(0x20, "Entity look", [
    ('eid', MC_int),
    ('yaw', MC_byte),
    ('pitch', MC_byte)])

srv_msgs[0x21] = defmsg(0x21, "Entity look/relative move", [
    ('eid',MC_int),
    ('dx',MC_byte),
    ('dy',MC_byte),
    ('dz',MC_byte),
    ('yaw',MC_byte),
    ('pitch',MC_byte)])

srv_msgs[0x22] = defmsg(0x22, "Entity teleport", [
    ('eid', MC_int),
    ('x', MC_int),
    ('y', MC_int),
    ('z', MC_int),
    ('yaw', MC_byte),
    ('pitch', MC_byte)])

srv_msgs[0x26] = defmsg(0x26, "Entity status", [
    ('eid',MC_int),
    ('status',MC_byte)])

cli_msgs[0x27] = \
srv_msgs[0x27] = defmsg(0x27, "Attach entity", [
    ('eid',MC_int),
    ('vehicle_id',MC_int)])

cli_msgs[0x28] = \
srv_msgs[0x28] = defmsg(0x28, "Entity metadata", [
    ('eid',MC_int),
    ('metadata',parse_metadata)])

cli_msgs[0x29] = \
srv_msgs[0x29] = defmsg(0x29, "Entity Effect", [
    ('eid', MC_int),
    ('effect_id', MC_byte),
    ('aplifier', MC_byte),
    ('duration', MC_short)])

cli_msgs[0x2a] = \
srv_msgs[0x2a] = defmsg(0x2a, "Remove entity effect", [
    ('eid', MC_int),
    ('effect_id', MC_byte)])

srv_msgs[0x2b] = defmsg(0x2b, "Experience", [
    ('curr_exp', MC_byte),
    ('level', MC_byte),
    ('tot_exp', MC_short)])

srv_msgs[0x32] = defmsg(0x32, "Pre-chunk", [
    ('x',MC_int),
    ('z',MC_int),
    ('mode',MC_bool)])

srv_msgs[0x33] = defmsg(0x33, "Chunk", [
    ('x',MC_int),
    ('y',MC_short),
    ('z',MC_int),
    ('size_x',MC_byte),
    ('size_y',MC_byte),
    ('size_z',MC_byte),
    ('chunk',MC_chunk)])

cli_msgs[0x34] = \
srv_msgs[0x34] = defmsg(0x34, "Multi-block change", [
    ('chunk_x',MC_int),
    ('chunk_z',MC_int),
    ('changes',MC_multi_block_change)])

cli_msgs[0x35] = \
srv_msgs[0x35] = defmsg(0x35, "Block change", [
    ('x',MC_int),
    ('y',MC_byte),
    ('z',MC_int),
    ('block_type',MC_byte),
    ('block_metadata',MC_byte)])

srv_msgs[0x36] = defmsg(0x36, "Play note block",[
    ('x', MC_int),
    ('y', MC_short),
    ('z', MC_int),
    ('instrument_type', MC_byte),
    ('pitch', MC_byte)])

srv_msgs[0x3c] = defmsg(0x3c, "Explosion", [
    ('x', MC_double),
    ('y', MC_double),
    ('z', MC_double),
    ('unknown', MC_float),
    ('records', MC_explosion_records)])

cli_msgs[0x46] = \
srv_msgs[0x46] = defmsg(0x46, "New/Invalid State", [
    ('reason', MC_byte),
    ('game_mode', MC_byte)])

srv_msgs[0x47] = defmsg(0x47, "Weather", [
    ('eid', MC_int),
    ('raining', MC_bool),
    ('x', MC_int),
    ('y', MC_int),
    ('z', MC_int)])

srv_msgs[0x64] = defmsg(0x64, "Open window", [
    ('window_id', MC_byte),
    ('inv_type', MC_byte),
    ('window_title', MC_string),
    ('num_slots', MC_byte)])

cli_msgs[0x65] = \
srv_msgs[0x65] = defmsg(0x65, "Close window", [
    ('window_id', MC_byte)])

cli_msgs[0x66] = defmsg(0x66, "Window click", [
    ('window_id', MC_byte),
    ('slot', MC_short),
    ('is_right_click', MC_bool),
    ('action_num', MC_short),
    ('shift', MC_bool),
    ('details', MC_item_details)])

srv_msgs[0x67] = defmsg(0x67, "Set slot", [
    ('window_id',MC_byte),
    ('slot',MC_short),
    ('slot_update',MC_slot_update)])

srv_msgs[0x68] = defmsg(0x68, "Window items", [
    ('window_id',MC_byte),
    ('inventory',MC_inventory)])

srv_msgs[0x69] = defmsg(0x69, "Update progress bar", [
    ('window_id', MC_byte),
    ('progress_bar',MC_short),
    ('value',MC_short)])

cli_msgs[0x6a] = \
srv_msgs[0x6a] = defmsg(0x6a, "Transaction", [
    ('window_id', MC_byte),
    ('action_num', MC_short),
    ('accepted', MC_bool)])

srv_msgs[0x6b] = defmsg(0x6b, "Creative inventory action", [
    ('slot', MC_short),
    ('item_id', MC_short),
    ('quantity', MC_short),
    ('damage', MC_short)])

cli_msgs[0x82] = \
srv_msgs[0x82] = defmsg(0x82, "Update sign", [
    ('x', MC_int),
    ('y', MC_short),
    ('z', MC_int),
    ('text1', MC_string),
    ('text2', MC_string),
    ('text3', MC_string),
    ('text4', MC_string)])

cli_msgs[0x83] = \
srv_msgs[0x83] = defmsg(0x83, "Item data", [
    ('item_type', MC_short),
    ('item_id', MC_short),
    ('data', MC_item_data)])

srv_msgs[0xc8] = defmsg(0xc8, "Increment statistic", [
    ('stat_id', MC_int),
    ('amount', MC_byte)])

srv_msgs[0xc9] = defmsg(0xc9, "Player list item", [
    ('name', MC_string),
    ('online', MC_bool),
    ('ping', MC_short)])

cli_msgs[0xfe] = defmsg(0xfe, "Server List Ping", [])

cli_msgs[0xff] = \
srv_msgs[0xff] = defmsg(0xff, "Disconnect/Kick", [
    ('reason', MC_string)])


